File: api/src/main/scala/org/eknet/county/Granularity.scala
package org.eknet.county

/**
 * @author Eike Kettner eike.kettner@gmail.com
 * @since 23.03.13 12:30
 */
sealed trait Granularity extends Serializable {

  def name: String

  def keyFor(millis: Long): TimeKey

}

object Granularity {

  val values = List(Millis, Second, Minute, Hour, Day, Month, Year)

  @SerialVersionUID(20130322L)
  object Millis extends Granularity {
    val name = "millis"
    def keyFor(millis: Long) = TimeKey(millis)
  }

  @SerialVersionUID(20130322L)
  object Second extends Granularity {
    val name = "second"
    def keyFor(millis: Long) = TimeKey(millis).bySeconds
  }

  @SerialVersionUID(20130322L)
  object Minute extends Granularity {
    val name = "minute"
    def keyFor(millis: Long) = TimeKey(millis).byMinutes
  }

  @SerialVersionUID(20130322L)
  object Hour extends Granularity {
    val name = "hour"
    def keyFor(millis: Long) = TimeKey(millis).byHour
  }

  @SerialVersionUID(20130322L)
  object Day extends Granularity {
    val name = "day"
    def keyFor(millis: Long) = TimeKey(millis).byDay
  }

  @SerialVersionUID(20130322L)
  object Month extends Granularity {
    val name = "month"
    def keyFor(millis: Long) = TimeKey(millis).byMonth
  }

  @SerialVersionUID(20130322L)
  object Year extends Granularity {
    val name = "year"
    def keyFor(millis: Long) = TimeKey(millis).byYear
  }
}

File: api/src/main/java/org/eknet/countyj/Granularity.java
/*
 * Copyright 2013 Eike Kettner
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.eknet.countyj;

import org.eknet.county.Granularity$;
import scala.collection.JavaConversions$;

/**
 * @author <a href="mailto:eike.kettner@gmail.com">Eike Kettner</a>
 * @since 25.03.13 19:28
 */
public enum Granularity {

  millis,
  second,
  minute,
  hour,
  day,
  month,
  year;

  public org.eknet.county.Granularity asScala() {
    for (org.eknet.county.Granularity g : JavaConversions$.MODULE$.asJavaCollection(Granularity$.MODULE$.values())) {
      if (g.name().equals(this.name())) {
        return g;
      }
    }
    throw new Error("unreachable code");
  }
}
